the-collab-lab/tcl-35-smart-shopping-list
# Aa dc sort list

## Description

This PR adds several functionalities:
- Sorting items by the estimated number of days until the next purchase.
- Sorting items with the same number of estimated days until next purchase alphabetically.
- Add background color to each item based on their `nextPurchase` days for visual distinction.
- Add `aria-label` to each item based on their `nextPurchase` days for accessibility purpose. 



## Related Issue

Closes #12 



## Acceptance Criteria

- Items in the list are shown as visually distinct (e.g., with a different background color on the list item) according to how soon the item is expected to be bought again: Soon, Kind of soon, Not soon, Inactive
- Items should be sorted by the estimated number of days until the next purchase
- Items with the same number of estimated days until the next purchase should be sorted alphabetically
- Items in the different states should be described distinctly when read by a screen reader



## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

![lis-item-before](https://user-images.githubusercontent.com/45172775/141113037-6eb08006-4757-4d27-8430-d4f2dd2ed9e7.jpg)




### After

![list-item](https://user-images.githubusercontent.com/45172775/141113085-d7cb635a-feea-454a-b965-02da9b46ceed.jpg)




## Testing Steps / QA Criteria

- Pull the branch.
- Run npm start
- Launch the browser to preview at http://localhost:3000/
- Enter a token from Firebase to join the existing list or create a new list and add several items.
  By now we should already see the background color for each item based on their `nextPurchase`.
   - Green -> 7 days
   - Yellow -> 14 days
   - Purple -> 30 days
- Open browser's devtools.
- Inspect the `<span>` inside the `<label>` of `input checkbox` to check the `aria-label`.
  Here we will see something like  `<span class="hide-span" aria-label="Buy Chocolate every 14 days">Chocolate</span>`.
- Go to Firebase.
- Rollback the date of `lastPurchase` on an item.
- Go to our application and check that item again.
  The expected behavior is that the list is now sorted based on the `estimatedPurchaseInterval`. And items with the same `estimatedPurchaseInterval` are sorted alphabetically.
